### PR TITLE
Improved GUI clipping with layers

### LINKED ISSUE #3117: Improve how GUI clipping works with layers (DEF-3906)
Would be great if we fix how GUI clipping nodes are working with the layers and in hierarchy, with layers, we would have the same draw order as in hierarchy without layers.

Old:
Using clipping and layers at the same time is confusing. Does it work according to what's in our manuals? Investigate!
2019-04-11 09:33:38 (Erik.Angelin)
(LINK REMOVED)
2019-05-02 16:43:31 (alexey.gulev)
According to our documentation, (LINK REMOVED)But now we have a difference:
 !Screenshot 2019-05-02 at 16.37.59.png|thumbnail!
the hierarchy with layer and the same hierarchy without layer works differently.

We need to look into the code to understand is it the bug or some lack of information in manuals.

Test project  [^DEF-3906-Layers-and-clipping.zip] 
2019-05-07 17:10:28 (jhonny.goransson)
I think the code is working as intended according to the manual: "Layers only affect the draw order of graphics—and furthermore, the layer set on a clipping node only affects the draw order in that clipper’s hierarchy."

All GUI render objects (box, particlefx etc) are sorted by the gui system by a render key. The render key is a uint64 split into five parts (from MSB to LSB):

[LAYER|INDEX|INV_CLIPPER_ID|SUB_LAYER|SUB_INDEX]

The bit count per field is not the same across these parts, but it doesn't matter. The render objects per frame are sorted in ascending order based on the uint64 key value. When the gui system traverses the tree to generate these render objects, it will generate a new 'clipping scope' for any root clipping nodes (i.e gui nodes that doesn't have a clipping parent in any part of its tree). A clipping scope contains three values; index, rootlayer and rootindex. The clipping scope will always set 0 in the layer field of the render key, and the actual node layer will be assigned to the sub_layer field. If there are several clipping nodes in a node sub-tree, the index of each clipping scope used will correspond to the level in that tree (I think, don't quote me on that).

This means that the layer property for a clipping node will only affect the render order of the nodes in that node hierarchy. Let's say you have a node A with layer 1 and node B with layer 1 AND clipping set to stencil. The render key for node A will be huge since the layer property will affect the most significant bits in the render key, whereas the render key for B will be significantly smaller and thus always rendered before A regardless of their relative ordering in the node tree.

A(layer 1)
 B(clipper, layer 1)
   - C

You would perhaps expect that the draw order is A,B,C, but in this case it will be B,C,A due to the render key calculation..

### PR DESCRIPTION
**Pull request by Jhonnyg**
The behaviour regarding how clipping interacts with layers in GUI scenes has been modified. Instead of treating each clipping node as its own isolated node tree, where the layers only affect the child nodes, using layers can now change the ordering of the nodes freely.

Fixes #3117

### CODE REVIEW

File: engine/gui/src/gui.cpp [hunk 1]
```diff
@@ -1284,20 +1275,23 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
...
-    static void Increment(Scope* scope) {
-        scope->m_Index = dmMath::Min(255, scope->m_Index + 1);
+    #if 0
+    static bool did_print = false;
+    void do_print_node(uint64_t name, uint64_t key)
+    {
+        if (did_print) return;
+        printf("%s, %lld\n", dmHashReverseSafe64(name), key);
     }
+    #endif
```
Comment (Jhonnyg) [inline engine/gui/src/gui.cpp:1285]: Remove before merge

File: engine/gui/src/gui.cpp [hunk 2]
```diff
@@ -295,16 +291,15 @@ namespace dmGui
...
+        params->m_MaxNodes       = 512;
+        params->m_MaxAnimations  = 128;
+        params->m_MaxTextures    = 32;
+        params->m_MaxFonts       = 4;
         params->m_MaxParticlefxs = 128;
-        // 16 is a hard cap for max layers since only 4 bits is available in the render key (see LAYER_RANGE above)
-        params->m_MaxLayers = 16;
+        // 256 is a hard cap for max layers, we use 8 bits in the render key (see LAYER_RANGE above)
+        params->m_MaxLayers       = 256;
         params->m_AdjustReference = dmGui::ADJUST_REFERENCE_LEGACY;
```
[the remark this answers is not in the document]
Comment (matgis) [reply engine/gui/src/gui.cpp:301]: Editor-note: There does not appear to be a corresponding limit in the editor, so there's nothing to update there.

File: editor/test/integration/gui_clipping_test.clj [hunk 3]
```diff
@@ -13,16 +13,11 @@
...
+            [clojure.test :refer :all]
             [dynamo.graph :as g]
-            [integration.test-util :as test-util]
-            [editor.workspace :as workspace]
-            [editor.defold-project :as project]
             [editor.gui :as gui]
-            [editor.gl.pass :as pass]
-            [editor.handler :as handler]
-            [editor.types :as types]))
+            [integration.test-util :as test-util]))
```
Comment (matgis) [inline editor/test/integration/gui_clipping_test.clj:20]: The changes to the editor tests should mirror the changes to the runtime tests in `test_gui_clipping.cpp`. However, the editor has a few more tests than the runtime for rendering, so there may be additional changes.

File: editor/test/integration/gui_clipping_test.clj [hunk 4]
```diff
@@ -745,7 +734,7 @@
       (add-layers! project scene ["layer1" "layer2"])
       (set-layer! a "layer2")
       (set-layer! c "layer1")
-      (assert-render-order scene [c d a]))))
+      (assert-render-order scene [c a d]))))
```
Comment (matgis) [inline editor/test/integration/gui_clipping_test.clj:737]: This editor test has no counterpart among the runtime tests.

File: editor/src/clj/editor/gui_clipping.clj [hunk 5]
```diff
@@ -589,51 +590,26 @@
...
-      scenes
-
-      (not (seq (first children-stack)))
-      (recur (rest children-stack) scenes)
-
-      true
-      (let [child (ffirst children-stack)
-            next-children (if (clipper-scene? child) [] (:children child))]
-        (recur (cons next-children (cons (rest (first children-stack)) (rest children-stack)))
-               (conj scenes child))))))
```
Comment (matgis) [inline editor/src/clj/editor/gui_clipping.clj:622]: After the change, clipper nodes are treated no differently than regular gui nodes, so this entire function can be replaced with a `tree-seq` call below.

**Review by vlaaad — APPROVED** [review]
LGTM!

File: engine/gui/src/gui.cpp [hunk 6]
```diff
@@ -1306,41 +1291,24 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
...
-                        uint64_t clipping_key = CalcRenderKey(current_scope, 0, 0);
-                        uint64_t render_key = CalcRenderKey(current_scope, layer, 1);
-                        CollectRenderEntries(scene, n->m_ChildHead, 2, current_scope, clippers, render_entries);
-                        if (layer > 0) {
-                            render_key = CalcRenderKey(current_scope, layer, 1);
-                        }
+                        uint64_t clipping_key = CalcRenderKey(layer, order++);
+                        uint64_t render_key   = CalcRenderKey(layer, order++);
+
+                        order = CollectRenderEntries(scene, n->m_ChildHead, order, clippers, render_entries);
```
Comment (JCash) [inline engine/gui/src/gui.cpp:1297]: I do like the cleanup!
A bit hard to reason about though but the recursive step seems ok I guess.
Also, don't forget to update the documentation!

File: engine/gui/src/test/test_gui_clipping.cpp [hunk 7]
```diff
@@ -774,7 +774,7 @@ TEST_F(dmGuiClippingTest, TestRenderOrder_ClipperStraight) {
  *   - b (layer1)
  * - c
  *
- * Expected order: a, b, c
+ * Expected order: a, c, b
```
Comment (JCash) [inline engine/gui/src/test/test_gui_clipping.cpp:777]: I guess I wouldn't have expected so many cases to change.
Are we expecting a big change for the users?

**Review by JCash — APPROVED** [review]

**Pull request merged**